File: Mr-Caesar/Day-6/max-product-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
 var maxProduct = function(nums) {
  let maxP = nums[0]
  let minP = nums[0]
  let res = nums[0]
  
  for(let j=1; j < nums.length; j++) {
      
      let x = Math.max(nums[j] * maxP, nums[j] * minP, nums[j])
      let y = Math.min(maxP * nums[j], nums[j] * minP, nums[j])    
      maxP = x
      minP = y        
      res = Math.max(res, maxP)            
  }
  return res
};

File: Richie/Maximum Product Subarray/max-product-subarray.py

class Solution:
    def maxSubArray(self, nums):
        mainMax = max(nums)
        currentMax = 1
        currentMin = 1
        
        for number in nums:
            tempVar = currentMax * number
            currentMax = max(number, currentMax * number, currentMin * number)
            currentMin = min(number, tempVar, currentMin * number)
            mainMax = max(mainMax, currentMax)

        return mainMax
    
numList = [2,3,-2,4]
print(Solution().maxSubArray(numList))
